Add in AddressGeometry entity to allow for answering questions such a…

seamuslee001 opened the pull request:
…s give me all geoemtries this address is in

The purpose of this is so that we can answer what geometries is an address in or what addresses are within geometry x

andrew-cormick-dockery commented:
I have a couple of questions:

- Is this effectively a denormalisation or a cache of geometries vs a particular address?  When does it get updated?  When a new geometry is added?  A new address?  Does this replace the DAG table?

- Why are you not using the namespace civigeometry?

seamuslee001 commented:
@andrew-cormick-dockery I would think that yes it might become a replacement for the dag, as for when this would be updated, I would imagine that may occur with changes to geometries or to addresses but i need to code up and think through the exact matrix. I'll fix the namespacing

seamuslee001 commented:
@andrew-cormick-dockery to answer your question as to when this is updated, i have now added in a post hook done in a symfony event listener to update it whenever an address is added / updated and whenever a geometry is archived. Also added tests now

andrew-cormick-dockery commented:
> 
> 
> @andrew-cormick-dockery to answer your question as to when this is updated, i have now added in a post hook done in a symfony event listener to update it whenever an address is added / updated and whenever a geometry is archived. Also added tests now

When a new geometry is added, I'm suspecting that this will be very CPU intensive.  I know that generating the DAG now is intensive, but still, how will this scenario be managed, do you intend?

andrew-cormick-dockery commented:
I should add that the listener for when a new address is added or geometry is archived is fine.  I'm just wondering about the scenario where we add a new branch, or we perform a redistribution.

seamuslee001 commented:
@andrew-cormick-dockery thanks for picking that up, just FYI John and i have been doing some testing and we haven't noticed any obvious CPU spike on dev when running a query like 

```
SELECT ca.id FROM civicrm_address ca, civigeometry_geometry g
WHERE g.id = 127
AND ST_CONTAINS(g.geometry, ST_GeomFromText(CONCAT('POINT(', ca.geo_code_2, ' ', ca.geo_code_1, ')'), 4326)) = 1
AND ca.is_primary = 1;
```
I just ran that exact query on production and it took 0.0946s . I would add that the current DAG is pretty non efficient right now

seamuslee001 commented:
@andrew-cormick-dockery i have now added code and a test to address that missing gap that you noticed
